iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/RedundantResourceDependenciesCheck.java

```diff
@@ -0,0 +1,169 @@
+import org.sonar.iac.arm.tree.api.StringLiteral;
+import org.sonar.iac.arm.tree.api.bicep.MemberExpression;
+import org.sonar.iac.common.api.checks.CheckContext;
+import org.sonar.iac.common.api.checks.IacCheck;
+import org.sonar.iac.common.api.checks.InitContext;
+import org.sonar.iac.common.api.checks.SecondaryLocation;
+import org.sonar.iac.common.api.tree.TextTree;
+import org.sonar.iac.common.checks.TextUtils;
+
+import static org.sonar.iac.arm.checks.utils.CheckUtils.isFunctionCall;
+
+@Rule(key = "S6952")
+public class RedundantResourceDependenciesCheck implements IacCheck {
+  private static final String MESSAGE = "Remove this explicit dependency as it is already defined implicitly.";
+  private static final String SECONDARY_MESSAGE_REFERENCE = "Implicit dependency is created via the \"reference\" function.";
+  private static final String SECONDARY_MESSAGE_SYMBOLIC = "Implicit dependency is created via a symbolic name.";
+  // Implicit dependency can be expressed only with `reference(s)?` functions:
+  // https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/template-functions-resource#implicit-dependency
+  // https://learn.microsoft.com/en-us/azure/azure-resource-manager/templates/template-functions-resource#implicit-dependency-1
+  private static final Predicate<Expression> IS_REFERENCE_FUNCTION = isFunctionCall("reference").or(isFunctionCall("references"));
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(ResourceDeclaration.class, RedundantResourceDependenciesCheck::collectExplicitDependencies);
+  }
+
+  private static void collectExplicitDependencies(CheckContext checkContext, ResourceDeclaration resourceDeclaration) {
+    var resourceExplicitDependencies = resourceDeclaration.getResourceProperty("dependsOn")
+      .map(Property::value)
+      .filter(ArrayExpression.class::isInstance)
+      .map(ArrayExpression.class::cast);
+    if (resourceExplicitDependencies.isEmpty()) {
+      return;
+    }
+
+    var explicitDependencies = collectDependencies(resourceExplicitDependencies.get());
+    var referencedResources = collectReferences(resourceDeclaration);
+    var referencedResourcesSymbolic = collectSymbolicReferences(resourceDeclaration);
+
+    for (var dependency : explicitDependencies) {
+      String dependencyName = dependency.value();
+
+      var references = new ArrayList<TextTree>();
+      for (var referencedResource : referencedResources) {
+        String referencedName = referencedResource.value();
+        if (dependencyName.equals(referencedName)) {
+          references.add(referencedResource);
+        }
+      }
+
+      var symbolicReferences = new ArrayList<TextTree>();
+      for (var referencedSymbol : referencedResourcesSymbolic) {
+        var referencedName = referencedSymbol.value();
+        if (dependencyName.equals(referencedName)) {
+          symbolicReferences.add(referencedSymbol);
+        }
+      }
```

Comment:
Those 2 loops can be rewritten using stream API - as the stream API is used in this check already a lot, it will be good to have consistency in the scope of one file (until there is a better reason not to use stream API, e.g., performance, but IMO it is not a case here).